In [30]:
import random
import numpy as np
import statistics

In [62]:
#Creating a class(A class is a blueprint of whatever we have to dob)
class KNN_Classifier():
  #Initiating the parameters
  def __init__(self,distance_metric):
    self.distance_metric = distance_metric      #Self takes the value of the instance we will create to predict later



  #Predicting the class of the test data
  def predict(self,X_train,test_data,k):
    neighbors=self.nearest_neighbors(X_train,test_data,k)

    for data in neighbors:
      label = []
      label.append(data[-1])

    predicted_class = statistics.mode(label)

    return predicted_class


  #Deciding the distance metric to be used and calculating distance based on this metric
  def get_distance_metric(self,train_data,X_test):       #Finding the Euclidean Distance
    if (self.distance_metric == 'euclidean'):
      distance = 0
      for i in range(len(train_data)-1):
        distance = distance+(train_data[i]-X_test[i])**2
      euc_distance=np.sqrt(distance)

      return euc_distance

    elif(self.distance_metric == 'manhattan'):                           #Finding the Manhattan Distance
      distance = 0
      for i in range(len(train_data)-1):
        distance = distance+abs(train_data[i]-X_test[i])
      man_distance=distance
      
      return man_distance 


  #Getting the nearest neighbors
  def nearest_neighbors(self,train_data,X_test,k):
    distance_list=[]

    for training_data in train_data:
      distance = self.get_distance_metric(training_data,X_test)
      distance_list.append((training_data, distance))
    
    distance_list.sort(key=lambda x:x[1])

    neighbors_list=[]

    for j in range(k):
      neighbors_list.append(distance_list[j][0])

    return neighbors_list





  

In [63]:
#Loading the KNN_Classifier to an instance 'classifer' for which .self was used
classifier = KNN_Classifier(distance_metric='euclidean')

In [33]:
import pandas as pd

churn = pd.read_csv('https://raw.githubusercontent.com/Halaarav/Some-Projects/main/churn.csv')

In [34]:
churn = churn.drop(columns=['RowNumber','CustomerId','Surname','Geography','Gender','Age'])

In [13]:
from sklearn.model_selection import train_test_split

In [35]:
X = churn.drop('Exited',axis=1)
y=churn.Exited
#Converting the data to numpy array
X = X.to_numpy()
y = y.to_numpy()


In [36]:
print(X),print(y)

[[6.1900000e+02 2.0000000e+00 0.0000000e+00 ... 1.0000000e+00
  1.0000000e+00 1.0134888e+05]
 [6.0800000e+02 1.0000000e+00 8.3807860e+04 ... 0.0000000e+00
  1.0000000e+00 1.1254258e+05]
 [5.0200000e+02 8.0000000e+00 1.5966080e+05 ... 1.0000000e+00
  0.0000000e+00 1.1393157e+05]
 ...
 [7.0900000e+02 7.0000000e+00 0.0000000e+00 ... 0.0000000e+00
  1.0000000e+00 4.2085580e+04]
 [7.7200000e+02 3.0000000e+00 7.5075310e+04 ... 1.0000000e+00
  0.0000000e+00 9.2888520e+04]
 [7.9200000e+02 4.0000000e+00 1.3014279e+05 ... 1.0000000e+00
  0.0000000e+00 3.8190780e+04]]
[1 0 1 ... 1 1 0]


(None, None)

In [37]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.25)

In [38]:
print(X_train.shape,X_test.shape,y_train.shape,y_test.shape)

(7500, 7) (2500, 7) (7500,) (2500,)


In [49]:
#Preparing the training data which will consist of all the features including the target variable whereas the test data will only contain the predictors
train_data = np.insert(X_train,7,y_train,axis=1)

In [74]:
X_test_size=X_test.shape[0]

In [77]:
prediction=[]
for i in range(X_test_size):
  prediction.append(classifier.predict(train_data,X_test[i],k=5))

In [79]:
from sklearn.metrics import accuracy_score

In [81]:
accuracy = accuracy_score(y_test,prediction)
accuracy

0.68

> Et voila! We. got a 68% accuracy for our custom KNN Classifier.\
There is a need for improvement and we shall look forward to it soon.